From a33a843758300644a0c3c0cf517249252c73c41c Mon Sep 17 00:00:00 2001
From: Ondrej Holy <oholy@redhat.com>
Date: Fri, 11 Aug 2017 13:05:27 +0200
Subject: [PATCH 07/13] channel: Set sockets as nonblocking to prevent
 deadlocks when copying

The channel socket pair is not set as nonblocking currently, which may cause
deadlocks in some cases (e.g. in-mount copy over read-write fallback), because
g_output_stream_write_async may block. This issue appears after increasing max
size of buffer in read channel:
https://bugzilla.gnome.org/show_bug.cgi?id=773826

Set channel sockets as nonblocking to be sure that _async methods don't block.

https://bugzilla.gnome.org/show_bug.cgi?id=785391
---
 daemon/gvfschannel.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/daemon/gvfschannel.c b/daemon/gvfschannel.c
index 021d292a..c3586603 100644
--- a/daemon/gvfschannel.c
+++ b/daemon/gvfschannel.c
@@ -209,7 +209,11 @@ g_vfs_channel_init (GVfsChannel *channel)
       channel->priv->cancellable = g_cancellable_new ();
       channel->priv->reply_stream = g_unix_output_stream_new (socket_fds[0], FALSE);
       channel->priv->remote_fd = socket_fds[1];
-      
+
+      /* Set as nonblocking to be sure that _async methods don't block. */
+      fcntl (socket_fds[0], F_SETFL, O_NONBLOCK);
+      fcntl (socket_fds[1], F_SETFL, O_NONBLOCK);
+
       start_request_reader (channel);
     }
 }
-- 
2.15.1

